packages/flutter_tools/lib/runner.dart

```diff
@@ -149,11 +150,41 @@ Future<int> _handleToolError(
     globals.printError('Oops; flutter has exited unexpectedly: "$error".');
 
     try {
+      String doctorText;
+      String piiStrippedDoctorText;
+      try {
+        final BufferLogger logger = BufferLogger(
+          terminal: globals.terminal,
+          outputPreferences: globals.outputPreferences,
+        );
+
+        // Run doctor twice: once with PII for the locally saved crash report, and again without PII
+        // for the GitHub template. The GitHub template is sent to the GitHub URL shortener service,
+        // avoid leaking PII.
+        final Doctor doctor = Doctor(logger: logger);
+
+        // Start the validator tasks only once.
+        final List<ValidatorTask> validatorTasks = doctor.startValidatorTasks();
+
+        // Run with PII.
+        await doctor.diagnose(showColor: false, startedValidatorTasks: validatorTasks);
+        doctorText = logger.statusText;
+
+        logger.clear();
+
+        // Run without PII. Do not send the doctor event a second time.
+        await doctor.diagnose(showColor: false, startedValidatorTasks: validatorTasks, showPii: false, sendEvent: false);
+        piiStrippedDoctorText = logger.statusText;
+      } on Exception catch (error, trace) {
+        doctorText = 'encountered exception: $error\n\n${trace.toString().trim()}\n';
+      }
```

Comment:
I moved all the `diagnose`, buffer clearing, and event sending logic to the new class as well, which also made it easier to test since it's not in the bowels of `runner`.